src/dbus/methods/shop_offers.c

```diff
@@ -0,0 +1,133 @@
+
+    repair_item_builder = g_variant_builder_new(G_VARIANT_TYPE("a{sv}"));
+    g_variant_builder_add(repair_item_builder, "{sv}", "name", g_variant_new_string(rci->name));
+    g_variant_builder_add(repair_item_builder, "{sv}", "repair_cost", g_variant_new_int32(rci->repair_cost));
+    g_variant_builder_add(repair_item_builder, "{sv}", "durability", g_variant_new_int32(rci->durability));
+
+    g_variant_builder_add(builder, "a{sv}", repair_item_builder);
+
+    g_variant_builder_unref(repair_item_builder);
+}
+
+static void olist_to_array(struct shop_offer *so, GVariantBuilder *builder)
+{
+    GVariantBuilder *item_builder, *repairs_builder;
+    GVariant *repairs_dict_array;
+
+    struct list *rl = so->repairs;
+
+    item_builder = g_variant_builder_new(G_VARIANT_TYPE("a{sv}"));
+
+    g_variant_builder_add(item_builder, "{sv}", "id", g_variant_new_int32(so->id));
+    g_variant_builder_add(item_builder, "{sv}", "name", g_variant_new_string(so->name));
+
+    g_variant_builder_add(item_builder, "{sv}", "supplier_id", g_variant_new_int32(so->supplier_id));
+    g_variant_builder_add(item_builder, "{sv}", "quantity", g_variant_new_int32(so->quantity));
+
+    g_variant_builder_add(item_builder, "{sv}", "expirationTime", g_variant_new_string(so->expirationTime));
+    g_variant_builder_add(item_builder, "{sv}", "durabilityPoints", g_variant_new_int32(so->durabilityPoints));
+    g_variant_builder_add(item_builder, "{sv}", "discount_percent", g_variant_new_int32(so->discount_percent));
+    g_variant_builder_add(item_builder, "{sv}", "rank", g_variant_new_int32(so->rank));
+
+    g_variant_builder_add(item_builder, "{sv}", "game_price", g_variant_new_int32(so->price.game.curr));
+    g_variant_builder_add(item_builder, "{sv}", "game_price_origin", g_variant_new_int32(so->price.game.orig));
+
+    g_variant_builder_add(item_builder, "{sv}", "cry_price", g_variant_new_int32(so->price.cry.curr));
+    g_variant_builder_add(item_builder, "{sv}", "cry_price_origin", g_variant_new_int32(so->price.cry.curr));
+
+    g_variant_builder_add(item_builder, "{sv}", "crown_price", g_variant_new_int32(so->price.crown.curr));
+    g_variant_builder_add(item_builder, "{sv}", "crown_price_origin", g_variant_new_int32(so->price.crown.curr));
+
+    g_variant_builder_add(item_builder, "{sv}", "current_price", g_variant_new_int32(so->price.key.curr));
+
+    repairs_builder = g_variant_builder_new(G_VARIANT_TYPE("aa{sv}"));
+    list_foreach(rl, (f_list_callback) rlist_to_array, repairs_builder);
+    repairs_dict_array = g_variant_builder_end(repairs_builder);
+
+    g_variant_builder_add(item_builder, "{sv}", "repairs", repairs_dict_array);
+
+    g_variant_builder_add(builder, "a{sv}", item_builder);
+
+    g_variant_builder_unref(item_builder);
+}
+
+void cmd_shop_offers_dbus_cb(const struct querycache *cache,
+                                  void *args)
+{
+    struct cb_args *a = (struct cb_args *) args;
+    struct list *so = session.wf.shop.offers;
```

Comment:
Even if the use of `session.wf.shop.offers` is correct here, I would instead prefer you to use the argument `cache` given in parameter of the function, like this `cache->container`  and `cache->hash`.